docs2/site/docs/guides/complexity-analyzer.md

```diff
@@ -0,0 +1,346 @@
+
+### 4. Logging Complexity Results
+
+In addition to validation, the `ValidateComplexityDelegate` property allows you to log complexity results
+for monitoring or analysis. Please note that this delegate runs after the maximum depth and complexity checks,
+so it cannot be used to log queries that fail these checks.
+
+```csharp
+complexityConfig.ValidateComplexityDelegate = async (context, complexity, depth) =>
+{
+    // RequestServices may be used to access scoped services within the DI container
+    var logger = context.RequestServices!.GetRequiredService<ILogger<MySchema>>();
+    logger.LogInformation($"Query Complexity: {complexity}, Depth: {depth}");
+};
+```
+
+To log failed complexity checks, set the `MaxDepth` and `MaxComplexity` properties to `null` and provide
+your own validation delegate:
+
+```csharp
+complexityConfig.MaxDepth = null;
+complexityConfig.MaxComplexity = null;
+complexityConfig.ValidateComplexityDelegate = async (context, complexity, depth) =>
+{
+    if (complexity > 1000 || depth > 10)
+    {
+        var logger = context.RequestServices!.GetRequiredService<ILogger<MySchema>>();
+        logger.LogWarning($"Query Complexity: {complexity}, Depth: {depth}");
+    }
+};
+```
+
+### 5. Throttling Users Based on Complexity Analysis
+
+To throttle users on a per-user basis similar to GitHub's GraphQL API limits, configure the
+complexity analyzer with a custom validation delegate. As noted above, `MaxComplexity` and `MaxDepth`,
+if set, are still enforced before this delegate runs.
+
+```csharp
+complexityConfig.ValidateComplexityDelegate = async (context, complexity, depth) =>
+{
+    // Get the authenticated user, or use the IP address if unauthenticated
+    var user = context.User;
+    string key;
+    if (user?.Identity?.IsAuthenticated == true)
+    {
+        // For authenticated users, use the user ID
+        key = "name:" + user.Identity.Name;
+    }
+    else
+    {
+        // For unauthenticated users, use the IP address
+        var httpContext = context.RequestServices!.GetRequiredService<IHttpContextAccessor>().HttpContext!;
+        key = "ip:" + httpContext.Connection.RemoteIpAddress.ToString();
+    }
+
+    // Pull your throttling service (e.g. Polly) from the DI container
+    var throttlingService = context.RequestServices!.GetRequiredService<IThrottlingService>();
+
+    // Throttle the request based on the complexity, subtracting the complexity from the user's limit
+    var (allow, remaining) = await throttlingService.ThrottleAsync(key, complexity);
+
+    // Report an error if the user has exceeded their limit
+    if (!allow)
+    {
+        context.ReportError(new ExecutionError($"Query complexity of {complexity} exceeded throttling limit. Remaining: {remaining}"));
+    }
+};
+```
+
+### 6. Throttling Users Based on Execution Time
+
+While the complexity analyzer does not directly measure execution time, you can write the following code
+to control the maximum execution time of a query:
```

Comment:
It looks unrelated to the topic. But it may be worth adding a dedicated property for the timeout configuration. And a dedicated docs page to mention different ways to configure the timeout: this property for simple config, `ConfigureExecution` for more complex scenarios, and APS.NET Core [middleware](https://learn.microsoft.com/en-us/aspnet/core/performance/timeouts?view=aspnetcore-8.0) when you don't need the GraphQL context.

Reply:
Where are you thinking a timeout would be added?  It's not really anything to do with the complexity configuration.  Is it in `ExecutionOptions`?  But then the timeout would only be able to wrap the actual execution, since the options will not be configured if we tried to wrap the entire chain.  I could add a PR and we can look at it.

Reply:
See:
- #3998

Reply:
Should we remove this section now?

Reply:
Updated